weave/type_wrappers/Content/content.py

```diff
@@ -0,0 +1,339 @@
+from __future__ import annotations
+
+import base64
+import hashlib
+import logging
+import os
+import subprocess
+import sys
+import uuid
+from pathlib import Path
+from typing import Annotated, Any, Generic, Literal, TypedDict, Union
+
+from pydantic import BaseModel, Field
+from typing_extensions import Self, TypeVar
+
+from .utils import default_filename, get_mime_and_extension, is_valid_b64, is_valid_path
+
+logger = logging.getLogger(__name__)
+
+# Dummy typevar to allow for passing mimetype/extension through annotated content
+# e.x. Content["pdf"] or Content["application/pdf"]
+T = TypeVar("T", bound=str)
+
+ContentType = Literal["bytes", "text", "base64", "file"]
+
+ValidContentInputs = Union[bytes, str, Path]
+
+
+# This is what is saved to the 'metadata.json' file by serialization layer
+# It is used to 'restore' an existing content object
+class ResolvedContentArgsWithoutData(TypedDict):
+    # Required Fields
+    id: str
+    size: int
+    mimetype: str
+    digest: str
+    filename: str
+    content_type: Literal["bytes", "text", "base64", "file", "url"]
+    input_type: str
+
+    # Optional Fields
+    extra: dict[str, Any] | None
+    path: str | None
+    extension: str | None
+    encoding: str | None
```

Comment:
For typed dict, optional fields are represented with `NotRequired`

Reply:
Messaged you on this.